cpu_scheduling_algorithms/round_robin_scheduling.cpp

```diff
@@ -0,0 +1,248 @@
+#include <iostream>   /// For outputting process execution results
+#include <queue>      /// Container for process execution turn
+#include <set>        /// Container for processes that have arrived
+#include <string>     /// For converting int type to string
+#include <utility>    /// For std::pair
+#include <vector>     /// Container for processes that will be executed
+
+/**
+ * @brief Represent a process to be executed.
+ */
+struct Process {
+    uint32_t id;            ///< Used to distinguish processes
+    uint32_t arrival_time;  ///< The time at which the process arrives
+    uint32_t burst_time;    ///< Time required to complete process execution
+};
+
+/**
+ * @brief Represent the result of a process execution.
+ */
+struct ProcessResult : public Process {
+    uint32_t completion_time;   ///< The time at which the process execution is
+                                ///< finished
+    uint32_t turn_around_time;  ///< The turn around time required for the
+                                ///< process to complete
+    uint32_t waiting_time;      ///< Process waiting time before execution
+
+    /**
+     * @brief Constructor that calculate member variables based on a
+     * process and completion time.
+     *
+     * \param process A process that have been executed
+     * \param completion_time The process execution finish time
+     */
+    ProcessResult(const Process& process, uint32_t completion_time)
+        : Process(process), completion_time(completion_time) {
+        turn_around_time = completion_time - arrival_time;
+        waiting_time = turn_around_time - burst_time;
+    }
+
+    /**
+     * @brief Compare each member variable.
+     *
+     * \param p ProcessResult to be compared with
+     * \return true if the processes IS equal
+     * \return false if the processes IS NOT equal
+     */
+    bool operator==(const ProcessResult& p) const {
+        return id == p.id && arrival_time == p.arrival_time &&
+               burst_time == p.burst_time &&
+               completion_time == p.completion_time &&
+               turn_around_time == p.turn_around_time &&
+               waiting_time == p.waiting_time;
+    }
+};
+
+/**
+ * Represent remaining burst time of a process.
+ */
+using BTLeft = uint32_t;
+
+/**
+ * @brief Execute processes based on Round-robin algorithm.
+ *
+ * \param processes Processes to be executed
+ * \param time_slice Time slice for processes execution
+ * \return Results of each process execution
+ */
+std::vector<ProcessResult> RRExecute(const std::vector<Process>& processes,
+                                     uint32_t time_slice);
+
+/**
+ * @brief Print a table containing process results data.
+ *
+ * \return ostream inputted ostream
+ */
+std::ostream& operator<<(std::ostream& ostream,
+                         const std::vector<ProcessResult>& results);
+
+/**
+ * @brief Self-test implementations.
+ *
+ * \returns void
+ */
+static void Test();
```

Comment:
I'm not sure if registering two functions is causing the Doxygen CI to fail or something. Not sure a prototype is needed for this.